keanu-project/src/main/java/io/improbable/keanu/distributions/discrete/Categorical.java

```diff
@@ -18,44 +20,51 @@ private Categorical(Map<T, DoubleTensor> selectableValues) {
         this.selectableValues = selectableValues;
     }
 
-    public T sample(int[] shape, KeanuRandom random) {
-        double sumOfProbabilities = getSumOfProbabilities();
-        double p = random.nextDouble();
-        double sum = 0;
-
-        if (sumOfProbabilities == 0.0) {
+    public GenericTensor<T> sample(int[] shape, KeanuRandom random) {
```

Comment:
At a high level I understand that this is the generalisation of the case of sampling a scalar (what it used to be). In the old version,  the logic was fairly penetrable in terms of understanding. 

In this new version, it's hard to see why this works. Is there anything we can do with breaking up into more functions, variable names and comments to make this more understandable? I'll add some more specific questions that come to mind when reading below.